packages/jsii-pacmak/bin/jsii-pacmak.ts

```diff
@@ -78,192 +87,153 @@ import { VERSION_DESC } from '../lib/version';
       default: true
     })
     .version(VERSION_DESC)
+    .strict()
     .argv;
 
   logging.level = argv.verbose !== undefined ? argv.verbose : 0;
 
-  logging.debug('command line arguments:', argv);
+  // Default to 4 threads in case of concurrency, good enough for most situations
+  const limit = argv.concurrency ? pLimit(4) : pLimit(1);
+  const awaitAll = makeAwaitAll(limit);
 
-  const rootDir = path.resolve(process.cwd(), argv._[0] || '.');
-
-  const visited = new Set<string>();
-  await buildPackage(rootDir, true /* isRoot */, argv['force-subdirectory']);
+  logging.debug('command line arguments:', argv);
 
-  async function buildPackage(packageDir: string, isRoot: boolean, forceSubdirectory: boolean) {
-    if (visited.has(packageDir)) {
-      return; // already built
-    }
+  const timers = new Timers();
 
-    visited.add(packageDir);
+  const modulesToPackage = await findJsiiModules(argv._, argv.recurse);
+  logging.info(`Found ${modulesToPackage.length} modules to package`);
+  if (modulesToPackage.length === 0) {
+    logging.warn('Nothing to do');
+    return;
+  }
 
-    // read package.json and extract the "jsii" configuration from it.
-    const pkg = await fs.readJson(path.join(packageDir, 'package.json'));
-    if (!pkg.jsii || !pkg.jsii.outdir || !pkg.jsii.targets) {
-      if (isRoot) {
-        throw new Error(`Invalid "jsii" section in ${packageDir}. Expecting "outdir" and "targets"`);
-      } else {
-        return; // just move on, this is not a jsii package
-      }
+  if (argv.outdir) {
+    for (const module of modulesToPackage) {
+      module.outputDirectory = path.resolve(argv.outdir);
     }
 
-    // if --recurse is set, find dependency dirs and build them.
-    if (argv.recurse) {
-      for (const dep of Object.keys(pkg.dependencies || {})) {
-        const depDir = resolveDependencyDirectory(packageDir, dep);
-        /* eslint-disable no-await-in-loop */
-        await buildPackage(depDir, /* isRoot */ false, forceSubdirectory);
-        /* eslint-enable no-await-in-loop */
-      }
-    }
+  } else if (argv.npmignore) {
+    // if outdir is coming from package.json, verify it is excluded by .npmignore. if it is explicitly
+    // defined via --out, don't perform this verification.
+    await updateAllNpmIgnores(modulesToPackage);
+  }
 
-    // outdir is either by package.json/jsii.outdir (relative to package root) or via command line (relative to cwd)
-    const outDir = argv.outdir !== undefined ? path.resolve(process.cwd(), argv.outdir) : path.resolve(packageDir, pkg.jsii.outdir);
-    const targets = argv.targets || [...Object.keys(pkg.jsii.targets), 'js']; // "js" is an implicit target.
+  await timers.recordAsync('npm pack', () => {
+    logging.info('Packaging NPM bundles');
+    return awaitAll(modulesToPackage
+      .map(m => () => m.npmPack()));
+  });
```

Comment:
Would we not extract the logging bit & make this a little less syntactically heavy?